examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -0,0 +1,209 @@
+        this.instanceIdx = instanceIdx;
+        this.transactionalId = "Processor-" + instanceIdx;
+        // A unique transactional.id must be provided in order to properly use EOS.
+        producer = new Producer(outputTopic, true, transactionalId, true, -1, null).get();
+        // Consumer must be in read_committed mode, which means it won't be able to read uncommitted data.
+        consumer = new Consumer(inputTopic, consumerGroupId, READ_COMMITTED, -1, null).get();
+        this.latch = latch;
+    }
+
+    @Override
+    public void run() {
+        // Init transactions call should always happen first in order to clear zombie transactions from previous generation.
+        producer.initTransactions();
+
+        final AtomicLong messageRemaining = new AtomicLong(Long.MAX_VALUE);
+
+        // Under group mode, topic based subscription is sufficient as EOS apps are safe to cooperate transactionally after 2.5.
+        // Under standalone mode, user needs to manually assign the topic partitions and make sure the assignment is unique
+        // across the consumer group instances.
+        if (this.mode.equals("groupMode")) {
+            consumer.subscribe(Collections.singleton(inputTopic), new ConsumerRebalanceListener() {
+                @Override
+                public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    printWithTxnId("Revoked partition assignment to kick-off rebalancing: " + partitions);
+                }
+
+                @Override
+                public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                    printWithTxnId("Received partition assignment after rebalancing: " + partitions);
+                    messageRemaining.set(messagesRemaining(consumer));
+                }
+            });
+        } else {
+            // Do a range assignment of topic partitions.
+            List<TopicPartition> topicPartitions = new ArrayList<>();
+            int rangeSize = numPartitions / numInstances;
+            int startPartition = rangeSize * instanceIdx;
+            int endPartition = Math.min(numPartitions - 1, startPartition + rangeSize - 1);
+            for (int partition = startPartition; partition <= endPartition; partition++) {
+                topicPartitions.add(new TopicPartition(inputTopic, partition));
+            }
+
+            consumer.assign(topicPartitions);
+            printWithTxnId("Manually assign partitions: " + topicPartitions);
+        }
+
+        int messageProcessed = 0;
+        boolean abortPreviousTransaction = false;
+        while (messageRemaining.get() > 0) {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(200));
+            if (records.count() > 0) {
+                try {
+                    // Abort previous transaction if instructed.
+                    if (abortPreviousTransaction) {
+                        producer.abortTransaction();
```

Comment:
It's a bit unconventional to have abort logic at the start of the loop. I think what users would expect is something like this:
```java
try {
  producer.beginTransaction()
  producer.send(...)
  producer.sendOffsetsToTransaction(...)
  producer.commitTransaction()
} catch (Exception ) {
  producer.abortTransaction()
}
```